elrond-wasm/src/storage/mappers/linked_list_mapper.rs

```diff
@@ -155,197 +181,245 @@ where
-    pub fn front(&self) -> Option<T> {
-        self.get_value_option(self.get_info().front)
-    }
-
-    /// Provides a copy to the back element, or `None` if the list is
-    /// empty.
-    pub fn back(&self) -> Option<T> {
-        self.get_value_option(self.get_info().back)
-    }
-
-    /// Removes the last element from a list and returns it, or `None` if
-    /// it is empty.
-    ///
-    /// This operation should compute in *O*(1) time.
-    pub fn pop_back(&mut self) -> Option<T> {
-        self.remove_by_node_id(self.get_info().back)
-    }
+    pub fn remove_node(&mut self, node: &LinkedListNode<T>) {
+        let node_id = node.node_id;
 
-    /// Removes the first element and returns it, or `None` if the list is
-    /// empty.
-    ///
-    /// This operation should compute in *O*(1) time.
-    pub fn pop_front(&mut self) -> Option<T> {
-        self.remove_by_node_id(self.get_info().front)
-    }
-
-    /// Removes element with the given node id and returns it, or `None` if the list is
-    /// empty.
-    /// Note: has undefined behavior if there's no node with the given node id in the list
-    ///
-    /// This operation should compute in *O*(1) time.
-    pub(crate) fn remove_by_node_id(&mut self, node_id: u32) -> Option<T> {
-        if node_id == NULL_ENTRY {
-            return None;
+        if self.is_empty_node(node_id) {
+            return;
         }
-        let node = self.get_node(node_id);
 
         let mut info = self.get_info();
-        if node.previous == NULL_ENTRY {
-            info.front = node.next;
+        if node.prev_id == NULL_ENTRY {
+            info.front = node.next_id;
         } else {
-            let mut previous = self.get_node(node.previous);
-            previous.next = node.next;
-            self.set_node(node.previous, previous);
+            let mut previous = self.get_node(node.prev_id);
+            previous.next_id = node.next_id;
+            self.set_node(node.prev_id, &previous);
         }
 
-        if node.next == NULL_ENTRY {
-            info.back = node.previous;
+        if node.next_id == NULL_ENTRY {
+            info.back = node.prev_id;
         } else {
-            let mut next = self.get_node(node.next);
-            next.previous = node.previous;
-            self.set_node(node.next, next);
+            let mut next = self.get_node(node.next_id);
+            next.prev_id = node.prev_id;
+            self.set_node(node.next_id, &next);
         }
 
         self.clear_node(node_id);
-        let removed_value = self.get_value(node_id);
-        self.clear_value(node_id);
         info.len -= 1;
         self.set_info(info);
-        Some(removed_value)
     }
 
-    /// Provides a forward iterator.
+    pub fn remove_node_by_id(&mut self, node_id: u32) -> Option<T> {
+        if self.is_empty_node(node_id) {
+            return None;
+        }
+
+        let node = self.get_node(node_id);
+        self.remove_node(&node);
+        Some(node.value)
+    }
```

Comment:
For consistency with the `get_node_by_id` method, this should also return the full node.

```suggestion
    pub fn remove_node_by_id(&mut self, node_id: u32) -> Option<LinkedListNode<T>> {
        if self.is_empty_node(node_id) {
            return None;
        }

        let node = self.get_node(node_id);
        self.remove_node(&node);
        Some(node)
    }
```